tcms/telemetry/api.py

```diff
@@ -226,6 +226,42 @@ def test_case_health(query=None):
     return data
 
 
+@http_basic_auth_login_required
+@rpc_method(name="Testing.individual_test_case_health")
+def individual_test_case_health(query=None):
+
+    if query is None:
+        query = {}
+
+    res = {}
+    plan_id = 0
+    positive, negative, all_count = 0, 0, 0
+    for test_execution in TestExecution.objects.filter(**query).order_by(
+        "run__plan_id"
+    ):
+        status = test_execution.status
+
+        if status.weight > 0:
+            positive += 1
+        elif status.weight < 0:
+            negative += 1
+        all_count += 1
+
+        if test_execution.run.plan_id != plan_id:
```

Comment:
Aggregating by plan_id only works in the scenario where we filter by case_id and query information for a single test case.  However this doesn't work for the Test Run page where we'd like to query for multiple cases (even filter by run_id directly instead of the case_id) and where information needs to be organized by case_id or TE.id. 

Maybe something like:

```
>>> for obj in TestExecution.objects.filter().values('run__plan', 'case_id', 'status__name', 'status__weight').order_by('case', 'run__plan', 'status__weight'):
...     print(obj)
... 
{'run__plan': 1, 'case_id': 1, 'status__name': 'ERROR', 'status__weight': -20}
{'run__plan': 1, 'case_id': 1, 'status__name': 'RUNNING', 'status__weight': 0}
{'run__plan': 1, 'case_id': 1, 'status__name': 'WAIVED', 'status__weight': 10}
{'run__plan': 1, 'case_id': 1, 'status__name': 'PASSED', 'status__weight': 20}
{'run__plan': 2, 'case_id': 1, 'status__name': 'PASSED', 'status__weight': 20}
{'run__plan': 1, 'case_id': 2, 'status__name': 'BLOCKED', 'status__weight': -10}
{'run__plan': 1, 'case_id': 2, 'status__name': 'RUNNING', 'status__weight': 0}
```

would be better on the API layer + an additional pre-processing layer on the FE depending on how we want to consume this information.

Reply:
@atodorov I pushed the changes to the API. however, I think the API is too generic this way, and we would need to do a lot of processing on the FE side. A better idea might be to have multiple APIs for the multiple use-cases of this feature, e.g. `Testing.individual_test_case_health.test_run` and `Testing.individual_test_case_health.test_case` for the different pages

Reply:
Indeed, the FE will need more processing. I think I've got a nice way to implement this into 1 API call.

For the TestCase page the query will be:  
Select from TE, Where case_id=X, plan_id IN [...], order & aggregate statuses by plan_id, 

where case_id is a scalar, plan_id is a list of IDs. The FE code on the TestCase page needs data[plan_id]['completion_rate | failure_rate']

For the TestRun page the query will be:
Select from TE, Where run_id=X, case_id IN[...], order & aggregate statuses by case_id

where run_id is a scalar, case_id is a list of IDs. The FE code here will need data[case_id]['completion_rate | failure_rate'].

I think if you modify the signature and accept 2 parameters while the 2nd one also controls the list of returned values and how they are aggregated we can have 1 generic function which will be consumed in 2 different places.

The inner workings of the function will not change but how it groups the results can be flexible depending on what we need.